### PR TITLE
Jetpack Cloud: Dismiss site selector on Escape or outside click

### PR DESCRIPTION
**Pull request by mattgawarecki**
This pull request adds event handlers to the page such that when someone opens the sidebar's site selector, they're able to dismiss it by clicking elsewhere on the page or by pressing the Escape key.

Resolves https://github.com/Automattic/jetpack-genesis/issues/87.

## Proposed changes

- Wrap `useOutsideClickCallback` event handler registrations in `setTimeout` to address a troublesome race condition on the Plugins page.
- Add a forward ref to the `SiteSelector` component so it can be referenced by other components (and dismissed, in our case).
- Refactor the site selector visibility toggle in the sidebar header to rely on app state, not local state.
- Implement a new site selector for Jetpack Cloud and use it in place of the original one.

## Testing instructions

- In your Jetpack Cloud testing environment, visit all pages, in both All Sites and single-site view. Do this in mobile and desktop viewports.
- On each page, verify the profile dropdown still functions as expected when clicked.
- On each page, verify that you're able to open and close the site selector in an intuitive way:
  - clicking on the site name in the sidebar should open the selector if it's not open
  - clicking outside the sidebar should dismiss the selector if it's open
  - pressing the Escape key should dismiss the selector if it's open
  - clicking or focusing the search bar in the site selector should work as expected (e.g., you can type, and the selector does not dismiss early or unexpectedly)

### CODE REVIEW

File: client/jetpack-cloud/components/sidebar/site-selector.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+		<BaseSiteSelector
+			forwardRef={ siteSelectorRef }
+			className="jetpack-cloud-sidebar__site-selector"
+			indicator
+			showAddNewSite
+			showAllSites
+			groups
+			/* eslint-disable-next-line jsx-a11y/no-autofocus */
+			autoFocus={ isVisible }
+			isJetpackAgencyDashboard={ canAccessJetpackManage }
```
Comment (yashwin) [inline client/jetpack-cloud/components/sidebar/site-selector.tsx:56]: This was `isJetpackAgencyDashboard={ isJetpackManage }` previously. Is this change intentional?

File: client/jetpack-cloud/components/sidebar/site-selector.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	return (
+		<BaseSiteSelector
+			forwardRef={ siteSelectorRef }
+			className="jetpack-cloud-sidebar__site-selector"
+			indicator
+			showAddNewSite
+			showAllSites
+			groups
+			/* eslint-disable-next-line jsx-a11y/no-autofocus */
+			autoFocus={ isVisible }
```
Comment (yashwin) [inline client/jetpack-cloud/components/sidebar/site-selector.tsx:55]: 👍

File: client/jetpack-cloud/components/sidebar/site-selector.tsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	const canAccessJetpackManage = useSelector( hasJetpackPartnerAccess );
+
+	return (
+		<BaseSiteSelector
+			forwardRef={ siteSelectorRef }
+			className="jetpack-cloud-sidebar__site-selector"
+			indicator
+			showAddNewSite
+			showAllSites
+			groups
```
Comment (yashwin) [inline client/jetpack-cloud/components/sidebar/site-selector.tsx]: Could we take a closer look at this? I haven't found where it's being used in this component except for `shouldShowGroups`, which is not used anywhere. Is it required for something that's not immediately apparent?

File: client/jetpack-cloud/components/sidebar/site-selector.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+			/* eslint-disable-next-line jsx-a11y/no-autofocus */
+			autoFocus={ isVisible }
+			isJetpackAgencyDashboard={ canAccessJetpackManage }
+			allSitesPath="/dashboard"
+			siteBasePath="/landing"
+			wpcomSiteBasePath="https://wordpress.com/home"
+			filter={ () => true }
```
Comment (yashwin) [inline client/jetpack-cloud/components/sidebar/site-selector.tsx]: I noticed this returns all sites. Perhaps we could consider removing it? What do you think?

File: client/jetpack-cloud/components/sidebar/site-selector.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	const siteSelectorRef = useHideSiteSelectorOnFocusOut();
+	const canAccessJetpackManage = useSelector( hasJetpackPartnerAccess );
+
+	return (
+		<BaseSiteSelector
+			forwardRef={ siteSelectorRef }
+			className="jetpack-cloud-sidebar__site-selector"
+			indicator
+			showAddNewSite
+			showAllSites
```
Comment (yashwin) [inline client/jetpack-cloud/components/sidebar/site-selector.tsx]: This was `showAllSites={ canAccessJetpackManage }` previously. Is this change intentional?

**Review by yashwin — COMMENTED** [review]
Thanks for making this change, @mattgawarecki! 

It works as expected. However, I have a few comments. Let me know what you think 🙂 

There is also a test case failure, not sure if that is related to this PR

File: client/jetpack-cloud/components/sidebar/site-selector.tsx (hunk 4 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/components/sidebar/site-selector.tsx]: I didn't check in my earlier run-through whether or not this field was safely okay not to define, but looking closer, it is. I agree, I'll remove this 👍

File: client/jetpack-cloud/components/sidebar/site-selector.tsx (hunk 3 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/components/sidebar/site-selector.tsx]: Agreed, will remove. I didn't examine whether or not properties were necessarily used when I wrote this; instead, I only copied over what was already working and tried to keep as close as possible to that baseline.

File: client/jetpack-cloud/components/sidebar/site-selector.tsx (hunk 1 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/components/sidebar/site-selector.tsx:56]: This change was intentional on my part, yeah -- `isJetpackAgencyDashboard` controls which "Add new site" button is displayed at the bottom of the selector, and if someone has access to Jetpack Manage we want them to be able to take advantage of its features, regardless of which part of Jetpack Cloud they're in at the moment. (Context: `isJetpackManage` is only set to true on pages that belong to Jetpack Manage itself, which excludes all the single-site pages like Backup, Scan, etc.)

File: client/jetpack-cloud/components/sidebar/site-selector.tsx (hunk 5 above)
Comment (mattgawarecki) [reply client/jetpack-cloud/components/sidebar/site-selector.tsx]: I think this was an oversight on my part, as if maybe I didn't circle back during my draft phase. I'm adding this back, because it only makes sense to show All Sites if the current user can access Jetpack Manage.

**Review by yashwin — APPROVED** [review]
LGTM. Thanks for addressing my comments.

It works as expected on large & small screen devices.

**Comment by yashwin** [comment]
However, there is a test case failure because of the below error:

Screenshot 2023-11-08 at 8 24 48 AM
